design-docs/opmodes-commandbased.md

```diff
@@ -0,0 +1,205 @@
+    opmode.running.whileTrue(Autos.followPath(this, path));
+  }
+
+  private void addArcadeTeleop() {
+    // A teleop opmode with joystick and button controls
+    CommandOpMode opmode = CommandOpModes.teleoperated("teleop");
+
+    var driverController = new CommandXboxController(1);
+
+    // Control the drive with split-stick arcade controls
+    m_drive.setDefaultCommand(
+        opmode,
+        drive.arcadeDriveCommand(
+            () -> -driverController.getLeftY(), () -> -driverController.getRightX()));
+
+    // Deploy the intake with the X button
+    opmode.running.and(driverController.x()).onTrue(intake.intakeCommand());
+    // Retract the intake with the Y button
+    opmode.running.and(driverController.y()).onTrue(intake.retractCommand());
+  }
+
+  @Override
+  public void beforeScheduler() {
+    // this code is called periodically in all robot modes of operation before the scheduler is run
+
+    // make sure we process vision inputs ahead of scheduled commands
+    vision.process();
+  }
+
+  @Override
+  public void afterScheduler() {
+    // this code is called periodically in all robot modes of operation after the scheduler is run
+
+    // output telemetry from all subsystems
+    Telemetry.log("drive", drive);
+    Telemetry.log("intake", intake);
+    Telemetry.log("storage", storage);
+  }
+}
+```
+
+Autos:
+
+```java
+public class Autos {
+  public static Command simpleAuto(Robot robot) {...}
+  public static Command followPath(Robot robot, String path) {...}
+}
+```
+
+# Drawbacks
+
+Command-based opmodes and non-command-based opmodes cannot be intermixed in a single project.  The benefit of this is cleaner integration with the rest of the command-based framework, including proper handling of periodic subsystem behaviors in disabled mode, which otherwise might be a common error for teams.  In addition, command-based subsystems would need to expose significant non-command-based behaviors to make them usable in non-command-based opmodes, which would likely create potential for additional issues in writing normal command-based code (where subsystem implementation details should be black-boxed from commands).
+
+# Alternatives
+
+Make command-based opmodes interoperate with non-command opmodes by using the same base class and only having the scheduler run if one of those opmodes is active.
+
+# Trades
+
+- Binding teleop joysticks is very verbose if different behavior is desired in different teleop opmodes.  Maybe add to CommandOpMode a separate event loop that's only active in that opmode?  At the minimum, it may make sense to add CommandOpMode overloads to joysticks so users could write `driverController.x(teleop)` instead of `teleop.running.and(driverController.x())`?
+- Is the `beforeScheduler()` / `afterScheduler()` split beneficial?  Or should we just have `robotPeriodic()` showing the `CommandScheduler.run()` code in it like we do in 2025?
```

Comment:
From my perspective, the `beforeScheduler()` / `afterScheduler()` split just creates extra functions. It's fairly clear what they are doing, but I don't think it's bad to keep the current approach with `CommandScheduler.run()` in `robotPeriodic()` is it is in 2025.

Reply:
I think if we do that, we need to clearly define whether the subsystem periodic() is called before or after the scheduler.  Right now it's not well defined and is a point of confusion for teams.